### DISCUSSION TITLE
Fix mypy errors in the entire codebase

### DISCUSSION THREAD

**Discussion by dhruvmanila** [TheAlgorithms/Python#8577] [Team Posts]
We need to fix all the `mypy` errors and they are increasing with the number of commits. I am planning to open an issue indicating this problem and asking for help. Some questions in my mind:

- As Python 3.9 is out, do we switch to `list[type]` or keep `typing.List`? Note: We do not yet run Python 3.9 on our CI as some of the dependencies aren't compatible yet.
- Some files do contain `from __future__ import annotation` and using `list[type]` but `mypy` complains about it: 
```
backtracking/minimax.py:15: error: "list" is not subscriptable, use "typing.List" instead
```

So, do we switch to Python 3.9 compatible type hints or advise users to user the `typing` module instead?

**Comment by cclauss** [comment]
At this point, we should switch to Python 3.9 (dropping any dependencies which are not ready) and then mypy should accept the syntax without the future import.

**Comment by dhruvmanila** [comment]
That would probably fail some of the checks as some files could rely on that dependency.

**Comment by xssbx** [comment]
I suggest a phased approach for Python 3.9 migration and mypy error fixes:
 
1. Retain Python 3.8+ compatibility first: Keep using  typing.List  instead of native  list[type]  for now, since CI doesn’t support 3.9 yet due to dependency issues. This fixes the mypy errors and keeps the codebase stable.
​
2. Prepare for 3.9 migration: Document all places using  typing.List  and create a task to switch to native type hints once dependencies are updated.
​
3. Handle  __future__ import annotations : This import is useful for Python 3.7+ type hinting, so keep it—resolve mypy’s dissatisfaction by adding the relevant ignore comments or updating the mypy config file.
